pkg/istio-agent/sds-agent.go

```diff
@@ -261,6 +261,74 @@ func gatewaySdsExists() bool {
 	return !os.IsNotExist(err)
 }
 
+// explicit code to determine the root CA to be configured in bootstrap file.
+// It may be different from the CA for the cert server - which is based on CA_ADDR
+// Replaces logic in the template:
+//                 {{- if .provisioned_cert }}
+//                  "filename": "{{(printf "%s%s" .provisioned_cert "/root-cert.pem") }}"
+//                  {{- else if eq .pilot_cert_provider "kubernetes" }}
+//                  "filename": "./var/run/secrets/kubernetes.io/serviceaccount/ca.crt"
+//                  {{- else if eq .pilot_cert_provider "istiod" }}
+//                  "filename": "./var/run/secrets/istio/root-cert.pem"
+//                  {{- end }}
+//
+// In addition it deals with the case the XDS server is on port 443, expected with a proper cert.
+// /etc/ssl/certs/ca-certificates.crt
+//
+// TODO: additional checks for existence. Fail early, instead of obscure envoy errors.
+func (sa *Agent) FindRootCAForXDS() string {
+	if strings.HasSuffix(sa.proxyConfig.DiscoveryAddress, ":443") {
+		return "/etc/ssl/certs/ca-certificates.crt"
+	} else if _, err := os.Stat("./etc/certs/root-cert.pem"); err == nil {
+		// Old style - mounted cert. This is used for XDS auth only,
+		// not connecting to CA_ADDR because this mode uses external
+		// agent (Secret refresh, etc)
+		return "./etc/certs/root-cert.pem"
+	} else if sa.secOpts.PilotCertProvider == "istiod" {
+		// PILOT_CERT_PROVIDER - default is istiod
+		// This is the default - a mounted config map on K8S
+		return "./var/run/secrets/istio/root-cert.pem"
+	} else if sa.secOpts.PilotCertProvider == "kubernetes" {
+		// Using K8S - this is likely incorrect, may work by accident.
+		// API is alpha.
```

Comment:
Removed. To be honest: I would remove the PilotCertProvider case completely, and just set ROOT_CA env in injector if it's k8s. Seems easier to undersatand.

The problem really is that in K8S it is not guaranteed that the cert is signed by that file - it's the CA for apiserver, not for the certs signed by k8s.  I don't know if beta added this guarantee, I should have mentioned that our code and integration is alpha - since we don't know the answer on 'who really signed the Istio certs'